### ISSUE TITLE
`updateHostHeader` doesn't update the host header when forwarding requests

### ISSUE THREAD

**Issue by niclim** [httptoolkit/mockttp#137]
Hi - I'm running something like...
```js
const server = mockttp.getLocal()

server
  .forAnyRequest()
  .always()
  .thenPassThrough({
    beforeRequest,
    forwarding: {
      targetHost: 'https://example.com',
      updateHostHeader: true,
    }
  })
```

This example does not update the host header for the forwarded request, however if I configure beforeRequest to modify the headers, it works (example below)
```js
const beforeRequest = (req) => {
  return { headers: { ...req.headers, host: newHost } }
}
```

Having a quick look through the code, it looks like `rawHeaders` is updated, but `headers` is not here https://github.com/httptoolkit/mockttp/blob/e769890bea6deb107ecc3c97114a77747fce6ba6/src/rules/requests/request-handlers.ts#L452-L458

**Comment by pimterry** [comment]
Ah, interesting! I suspect this is an interaction between `beforeRequests` and `updateHostHeader` - if I remember rightly using the former effectively skips some 'preserve raw header' logic (to make header handling within the callback more convenient) and I expect that means it uses the original non-raw headers, ignoring modifications to raw headers in the meantime.

If you remove the `beforeRequest` it should work correctly (but of course I imagine that doesn't work for your case, and it certainly shouldn't be required).

Want to put a PR together to fix this?

I suspect the right solution is to avoid using `headers` anywhere internally in this `handle()` method to always use/modify `rawHeaders` throughout instead, and whenever we need processed headers for an API (like `beforeRequest`) we should rebuild it from the current raw headers first. That way there's just a single source of truth for header data (much better than trying to keep both in sync all the time).

In theory we should already be doing something close to that, but clearly it that isn't correctly handled everywhere right now.

**Comment by niclim** [comment]
Sure, I can take a look and put together a PR sometime this week

**Comment by pimterry** [comment]
Released in v3.7.2. Thanks for contributing!

**Comment by niclim** [comment]
Thanks @pimterry! I was trying to upgrade and I think there's a bug that was introduced on this line

https://github.com/httptoolkit/mockttp/blob/8e2c24e0e4361b320152d17fdb349bea9f551101/src/rules/requests/request-handlers.ts#L473

Forwarding requests from a url results in a double slash `http://localhost:8000/users` -> `http://localhost:8001//users` since `url.parse().path` includes a leading slash

**Comment by pimterry** [comment]
Good catch @niclim! The value of `req.url` inside `beforeRequest` was another case I'd missed here - I've now added tests for that (which did indeed fail with that double slash issue) and fixed it. Thanks! Now released as v3.7.3.
